### DISCUSSION TITLE
best way to search contents of large file with small file

### DISCUSSION THREAD

**Discussion by Shooter3k** [BurntSushi/ripgrep#2089] [Q&A] [answered]
I've been searching for days and I haven't found a good way to search for 

my current command is basically this and regardless of which grep I use, nothing is executing this using hardly any cpu threads. In fact, the nvme drive I'm using this on doesn't even show it's getting access because it's so little. 

rg  -f 2mb_lookup_file 30gb_master_file 

Anyone have any ideas/suggestions?

**Comment by BurntSushi** [comment, marked as answer]
Because building what is likely thousands of regexes takes a ton of work and memory. But without any kind of reproduction, I can't really answer your question with confidence.

**Comment by Shooter3k** [comment]
The 2mb file and the 30gb file are lists of SHA1 hashes. The 2mb file is just the hash, while the 30gb file is a list of hashes:plain 

I'd be happy to share them but not sure you'd want a 30gb file?

**Reply by BurntSushi** [reply]
I think I would only need the 2MB file to confirm my guess.

Could you try using the `-F` flag?

**Reply by Shooter3k** [reply]
Using only the -F flag somehow magically finishes in 25 seconds after it reads the file at amazing speeds but it doesn't show me any results

**Reply by BurntSushi** [reply]
I assume you expect some results?

But yeah, please share the input patterns if you can.

`-F` in theory should not be required, but there is an optimization that kicks in to use Aho-Corasick to paper over limitations in the current implementation of the regex engine. Building using Aho-Corasick is much much faster than building a full regex. A lot less overhead.

**Reply by Shooter3k** [reply]
here's the file https://mega.nz/file/bwpEACCD#8KpBdtAwfk_f2YHDr53aBaRJuh5cgVLubXFu0tPcBgA

and then if you just take any one of those rows and put them in a second file and put :test at the end of the row, that's what the file looks like that I'm searching against

**Reply by BurntSushi** [reply]
Seems to be working fine for me:

```
$ time rg -F -f hashes.txt corpus.txt
1:bccb4505522a0a4d01c319380915d02d:test

real    1.606
user    1.438
sys     0.163
maxmem  872 MB
faults  0

$ time rg -f hashes.txt corpus.txt
1:bccb4505522a0a4d01c319380915d02d:test

real    2.618
user    2.239
sys     0.266
maxmem  1232 MB
faults  0

$ cat corpus.txt
bccb4505522a0a4d01c319380915d02d:test
```

Could you please provide a `corpus.txt` file that reproduces your issue where ripgrep doesn't print any results but should? It doesn't need to be 30GB.

Also, running `perf` does indeed confirm that most of the time is just being spent building the matching engine.

**Reply by Shooter3k** [reply]
I kept trying to reproduce my original issue and I kept getting this error: 
Binary file matches (found "\u{0}" byte around offset 0)

I edited the corpus file by deleting the first few lines of seemingly null characters and it's working now. 

So, the good news is that it's working. The bad news is that it still takes about 17-20 minutes to search the 15gb file when I use -F -f  or -f  but takes about 7 seconds when I use -F only. 

Do I have any other options besides using -F -f or -f? If it takes 17-20 minutes then I might just end up turning these files into database files and have to search them that way instead (which I'm really really trying to avoid).

**Reply by BurntSushi** [reply]
I can't answer you without more of a reproduction. Maybe share a smaller file where you can still see a performance difference. And please include precisely the commands you are running. For example, "not using -f" doesn't make sense to me. How can you not use -f when you're searching for tens of thousands of patterns?

> Do I have any other options besides using -F -f or -f?

Like what..?

I think it would help if you took a moment to reflect on what needs to happen to search for 100,000+ needles in a 15GB haystack. Give it some thought.

> database files 

It seems to me like you could write just a little code here. Read your patterns into a hashmap. Iterate over lines in corpus. Check if hash is in map.

Alternatively, sort the files and do a set intersection that way.

Maybe the best approach would be to put your corpus into an fst (see the fst crate) and then search that. Some databases use fsts internally as their index structures.

**Reply by Shooter3k** [reply]
here's a piece of the "big" file that should work to compare to the smaller file earlier https://mega.nz/file/bhBkQDxA#UkeU2KPFOxD2kK0NUEELk3xYURnJdCGJHepZIWOP99Q

time rg -F -f hashes.txt 0.testfile_split_ab.txt

time rg -f hashes.txt 0.testfile_split_ab.txt

I'm assuming your previous answers will be the same but if they are different; any suggestions on how to get this to run as fast as physically possible?
